nexus/src/app/rack.rs

```diff
@@ -219,6 +219,76 @@ impl super::Nexus {
             self.background_tasks.activate(task);
         }
 
+        // TODO - levon
+        // register all switches found during rack initialization
+        // identify requested switch from config and associate
+        // uplink records to that switch
+        match request.external_port_count {
+            ExternalPortDiscovery::Auto(switch_mgmt_addrs) => {
+                use dpd_client::Client as DpdClient;
+                info!(
+                    self.log,
+                    "Using automatic external switchport discovery"
+                );
+
+                for (switch, addr) in switch_mgmt_addrs {
+                    let dpd_client = DpdClient::new(
+                        &format!(
+                            "http://[{}]:{}",
+                            addr,
+                            omicron_common::address::DENDRITE_PORT
+                        ),
+                        dpd_client::ClientState {
+                            tag: "rss".to_string(),
+                            log: self.log.new(o!("component" => "DpdClient")),
+                        },
+                    );
+
+                    let all_ports =
+                        dpd_client.port_list().await.map_err(|e| {
+                            Error::internal_error(&format!("encountered error while discovering ports for {switch:#?}: {e}"))
+                        })?;
+
+                    info!(
+                        self.log,
+                        "discovered ports for {switch}: {all_ports:#?}"
+                    );
+
+                    let qsfp_ports: Vec<Name> = all_ports
+                        .iter()
+                        .filter(|port| port.starts_with("qsfp"))
+                        .map(|port| port.to_string().parse().unwrap())
+                        .collect();
+
+                    info!(
+                        self.log,
+                        "populating ports for {switch}: {qsfp_ports:#?}"
+                    );
+
+                    self.populate_switch_ports(
+                        &opctx,
+                        &qsfp_ports,
+                        switch.to_string().parse().unwrap(),
+                    )
+                    .await?;
+                }
+            }
+            ExternalPortDiscovery::Static(port_mappings) => {
```

Comment:
If we have automatic switch port discovery, what is the utility of keeping static around?

Reply:
Unit tests / integration tests pass a static value directly to `nexus.rack_initialize`. It might be possible to plumb in some logic to call the `dpd` instance that gets created in the test context, but at the time I chose to kick the can on in light of more pressing issues.

Reply:
https://github.com/oxidecomputer/omicron/issues/3602